supportedTransports.js

```diff
@@ -0,0 +1,105 @@
+};
+
+// Fallback order inspired by https://github.com/vacuumlabs/adalite
+async function isWebUsbSupported() {
+    try {
+        const isSupported = await LedgerTransportWebUsb.isSupported();
+        return isSupported && platform.os.family !== 'Windows' && platform.name !== 'Opera';
+    } catch (e) {
+        return false;
+    }
+}
+
+async function isWebHidSupported() {
+    return LedgerTransportWebHid.isSupported()
+        .catch(() => false);
+}
+
+async function isU2fSupported() {
+    return LedgerTransportWebHid.isSupported()
+        .catch(() => false);
+}
+
+async function createSupportedTransport() {
+    const [
+        supportWebHid,
+        supportWebUsb,
+        supportU2f
+    ] = await Promise.all([
+        isWebHidSupported(),
+        isWebUsbSupported(),
+        isU2fSupported()
+    ]);
+
+    debugLog("Transports supported:", { supportWebHid, supportWebUsb, supportU2f });
+
+    if (!supportWebHid && !supportWebUsb && !supportU2f) {
+        const err = new Error('No transports appear to be supported.');
+        err.name = 'NoTransportSupported';
+        throw err;
+    }
+
+    // Sometimes transports return true for `isSupported()`, but are proven broken when attempting to `create()` them.
+    // We will try each transport we think is supported in the current environment, in order of this array
+    const supportedTransports = [
+        // WebHID is still supported by latest Chrome, so try that first
+        ...(supportWebHid && [{ name: 'WebHID', createTransport: () => LedgerTransportWebHid.create() }]),
+
+        ...(supportWebHid && [{ name: 'WebUSB', createTransport: () => LedgerTransportWebUsb.create() }]),
+
+        // Firefox/Mozilla intend to not support WebHID or WebUSB
+        ...(supportWebHid && [{ name: 'U2F', createTransport: () => LedgerTransportU2F.create() }]),
+    ]
+
+    let transport = null;
+    let errors = [];
+
+    for (let i = 0; i < supportedTransports.length && !transport; i += 1) {
+        const { name, createTransport } = supportedTransports[i];
+        debugLog(`Creating ${name} transport`)
+        try {
+            transport = await createTransport();
+        } catch (err) {
+            // If the user clicked the `cancel` button, stop attempting fallbacks
+            if (err.name === 'TransportOpenUserCancelled') {
+                throw err;
+            }
+
+            console.warn(`Failed to create ${name} transport.`, err);
+            errors.push({ name: err.name, message: err.message });
+        }
+    }
+
+    return [errors, transport];
+}
+
+module.exports.setDebugLogging = (value) => ENABLE_DEBUG_LOGGING = value;
+module.exports.getSupportedTransport = async function getSupportedTransports() {
+    try {
+        const [errors, transport] = await createSupportedTransport();
+
+        if (errors && !transport) {
+            console.error('Failed to initialize ledger transport', { errors });
+            throw errors[errors.length - 1];
+        }
```

Comment:
what if there are no errors and no transport?

Reply:
In that case, `getSupportedTransports()` will return `null`, which is I think OK for the caller.
Now that you mention it, though, the log in `getSupportedTransport()` for 'transport created!' shouldn't happen if `transport` is null.  Fixed! :)